htlcswitch/link.go

```diff
@@ -609,47 +701,107 @@ func (l *channelLink) IsDraining(linkDirection LinkDirection) bool {
-	return fn.None[FlushHookID]()
+func (l *channelLink) OnFlushedOnce(hook func()) fn.Option[FlushHookID] {
+	// If the channel state is clean, meaning there are no HTLCs, and both
+	// peers have irrevocably committed that state, then we can call this
+	// hook synchronously and not add it to the map at all.
+	if l.channel.IsChannelClean() {
+		hook()
+		return fn.None[FlushHookID]()
+	}
+
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	return fn.Some(FlushHookID(l.flushHooks.alloc(hook, false)))
 }
 
 // OnFlushedMany is a method that adds a hook to be called every time the
 // channel state reaches zero htlcs. This function returns a FlushId that can be
 // used to preemptively remove this hook.
-func (l *channelLink) OnFlushedMany(func()) FlushHookID {
-	// TODO(proofofkeags): Implement
-	return FlushHookID(0)
+func (l *channelLink) OnFlushedMany(hook func()) FlushHookID {
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	return FlushHookID(l.flushHooks.alloc(hook, true))
 }
 
 // RemoveFlushHook will remove a hook previously added by OnFlushedOnce or
 // OnFlushedMany. It returns an error if the supplied FlushId was not found.
-func (l *channelLink) RemoveFlushHook(FlushHookID) error {
-	// TODO(proofofkeags): Implement
-	return errors.New("no flush in progress to cancel")
+func (l *channelLink) RemoveFlushHook(hid FlushHookID) error {
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	if l.flushHooks.delete(uint64(hid)) {
+		return nil
+	}
+
+	return errors.New("could not remove flush hook: not found")
 }
 
 // OnCommitOnce adds a hook that will be called the next time a CommitSig
 // message is sent in the argument's LinkDirection.
 func (l *channelLink) OnCommitOnce(
-	LinkDirection, func(),
+	direction LinkDirection, hook func(),
 ) fn.Option[CommitHookID] {
 
-	// TODO(proofofkeags): Implement
-	return fn.None[CommitHookID]()
+	// If we want to run the hook once on the next fully committed state,
+	// and we don't currently need/owe a commitment, it means we don't have
+	// any pending updates and we can run the hook synchronously. If we call
+	// it syncrhonously we will not add it to the map at all.
+	runSync := direction == Outgoing && !l.channel.OweCommitment() ||
+		direction == Incoming && !l.channel.NeedCommitment()
+	if runSync {
+		hook()
+		return fn.None[CommitHookID]()
+	}
+
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	var hm *hookMap
+	if direction == Outgoing {
+		hm = &l.outgoingCommitHooks
+	} else {
+		hm = &l.incomingCommitHooks
+	}
+
+	return fn.Some(CommitHookID(hm.alloc(hook, false)))
 }
 
 // OnCommitMany adds a hook that will be called every time a CommitSig message
 // is sent in the argument's LinkDirection.
-func (l *channelLink) OnCommitMany(LinkDirection, func()) CommitHookID {
-	// TODO(proofofkeags): Implement
-	return CommitHookID(0)
+func (l *channelLink) OnCommitMany(
+	direction LinkDirection, hook func(),
+) CommitHookID {
+
+	l.RWMutex.Lock()
+	defer l.RWMutex.Unlock()
+
+	var hm *hookMap
```

Comment:
Doesn't actually need to be a pointer?

Reply:
It does because we choose the value that goes in here off of the following block and the `alloc` method takes a pointer receiver. We don't want to call by value for `alloc` either.